upstairs/src/lib.rs

```diff
@@ -2607,6 +2388,212 @@ impl BlockIO for Guest {
+///
+/// * Wait for them to complete, then notify the guest through oneshot channels
+pub struct GuestIoHandle {
+    /// Queue to receive new blockreqs
+    req_rx: mpsc::Receiver<BlockReq>,
+
+    /// Guest IO and bandwidth limits
+    limits: Arc<std::sync::Mutex<GuestLimits>>,
+
+    /// `BlockReq` that is at the head of the queue
+    ///
+    /// If a `BlockReq` was pulled from the queue but couldn't be used due to
+    /// IOP or bandwidth limiting, it's stored here instead (and we check this
+    /// before awaiting the queue).
+    req_head: Option<BlockReq>,
+
+    /// Are we currently IOP or bandwidth limited?
+    ///
+    /// If so, we don't return anything in `recv()`
+    req_limited: bool,
+
+    /// Current number of IOP tokens
+    iop_tokens: usize,
+
+    /// Current backpressure (shared with the `Guest`)
+    backpressure_us: Arc<AtomicU64>,
+
+    /// Backpressure configuration, as a starting point and max delay
+    backpressure_config: BackpressureConfig,
+
+    /// Bandwidth tokens (in bytes)
+    bw_tokens: usize,
+
+    /// Active work from the guest
+    ///
+    /// When the crucible listening task has noticed a new IO request, it
+    /// will pull it from the reqs queue and create an GuestWork struct
+    /// as well as convert the new IO request into the matching
+    /// downstairs request(s). Each new GuestWork request will get a
+    /// unique gw_id, which is also the index for that operation into the
+    /// hashmap.
+    ///
+    /// It is during this process that data will encrypted. For a read, the
+    /// data is decrypted back to the guest provided buffer after all the
+    /// required downstairs operations are completed.
+    guest_work: GuestWork,
+
+    /// Log handle, mainly to pass it into the [`Upstairs`]
+    log: Logger,
+}
+
+impl GuestIoHandle {
+    /// Leak IOPs tokens
+    pub fn leak_iop_tokens(&mut self, tokens: usize) {
+        self.iop_tokens = self.iop_tokens.saturating_sub(tokens);
+        self.req_limited = false;
+    }
+
+    /// Leak bytes from bandwidth tokens
+    pub fn leak_bw_tokens(&mut self, bytes: usize) {
+        self.bw_tokens = self.bw_tokens.saturating_sub(bytes);
+        self.req_limited = false;
+    }
+
+    /// Listen for new work
+    ///
+    /// This will wait forever if we are currently IOP / BW limited; otherwise,
+    /// it will return the next value from the `BlockReq` queue.
+    async fn recv(&mut self) -> BlockReq {
+        let req = if self.req_limited {
+            futures::future::pending().await
+        } else if let Some(req) = self.req_head.take() {
+            req
+        } else if let Some(req) = self.req_rx.recv().await {
+            // NOTE: once we take this req from the queue, we must be cancel
+            // safe!  In other words, we cannot yield until either (1) returning
+            // the req or (2) storing it in self.req_head for safe-keeping.
+            req
+        } else {
+            futures::future::pending().await
```

Comment:
If we enter this branch, it means that the `recv` above returned `None` right? In that case the `req` channel was closed: it seems like `crucible_bail!(RecvDisconnected)` (or an equivalent for this new channel) may apply here?

Reply:
I'm not sure!

In the current code, there's nothing to acknowledge when the last clone of the `Arc<Guest>` (that _isn't_ owned by the upstairs) is destroyed; we just stop seeing messages come through from the one remaining clone.  This means that the `Upstairs` will pick a different action in its event loop, since we have plenty of remaining valid `select!` branches.

This code matches that existing behavior: if the `Guest` is dropped, we'll never see any more messages from it, and the `Upstairs` will go about its business by picking one of the other `select!` branches.

My inclination is to keep the current logic, but add a logged message to that effect.  Maybe in the future we'd want to quit the Upstairs event loop if the `Guest` is dropped, since no one would be able to control it at that point, but I think that could come in a later PR.

What do you think?

Reply:
> In the current code, there's nothing to acknowledge when the last clone of the Arc<Guest> (that isn't owned by the upstairs) is destroyed; we just stop seeing messages come through from the one remaining clone.

With this PR, it may be ok to abort if we see None here: `Upstairs` now holds `GuestIoHandle`, not `Guest`, so I think the last drop means the channel will be closed?

Reply:
I'm not sure I understand what you mean by "ok to abort".

The `Upstairs` holds the `GuestIoHandle`'s `mpsc::Receiver` end, but the `Guest` holds the `mpsc::Sender`.  If the `Guest` is dropped, then the receiver will see `None` here.  As such, if we panicked upon seeing `None`, we'd panic whenever the guest is dropped, which we don't have control over (since that's managed by whoever has embedded the `crucible-upstairs` library).

Here's a diff to double-check:
```diff
diff --git a/integration_tests/src/lib.rs b/integration_tests/src/lib.rs
index c767476..61967ab 100644
--- a/integration_tests/src/lib.rs
+++ b/integration_tests/src/lib.rs
@@ -3102,6 +3102,24 @@ mod test {
         Ok(())
     }

+    #[tokio::test]
+    async fn integration_test_guest_drop() -> Result<()> {
+        // Spin off three downstairs, build our Crucible struct.
+        let tds = TestDownstairsSet::small(false).await?;
+        let opts = tds.opts();
+
+        let (guest, io) = Guest::new(None);
+
+        let join_handle = up_main(opts, 1, None, io, None)?;
+
+        drop(guest);
+        tokio::time::sleep(std::time::Duration::from_secs(1)).await;
+
+        let r = join_handle.await;
+        assert!(r.is_ok());
+        Ok(())
+    }
+
     #[tokio::test]
     async fn integration_test_guest_zero_length_io() -> Result<()> {
         // Test the guest layer with a write and read of zero length
diff --git a/upstairs/src/lib.rs b/upstairs/src/lib.rs
index 8644305..f208408 100644
--- a/upstairs/src/lib.rs
+++ b/upstairs/src/lib.rs
@@ -2505,7 +2505,7 @@ impl GuestIoHandle {
         } else {
             if !self.guest_dropped {
                 self.guest_dropped = true;
-                warn!(self.log, "Guest handle has been dropped");
+                panic!("Guest handle has been dropped");
             }
             futures::future::pending().await
         };
```

This unit test fails because this panic fires when the `Guest` is dropped:
```
thread 'test::integration_test_guest_drop' panicked at upstairs/src/lib.rs:2508:17:
Guest handle has been dropped
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

...and then the join handle returns an error: `Err(JoinError::Panic(Id(13), ...))`

(This would be even more dramatic with `panic=abort`, which isn't enabled for unit tests)

Reply:
Got it, please take a look at 6874eb3b89f8d64f577714ba3bc094fc53ac5e0e (while we wait for CI to test everything)

Reply:
Nice :+1: